File: copy-supabase-to-local.js
const { Pool } = require('pg');

const supabasePool = new Pool({
  connectionString: 'postgresql://postgres.fojbzghphznbslqwurrm:fulebimojviT1985%@aws-1-eu-west-1.pooler.supabase.com:5432/postgres',
  ssl: { rejectUnauthorized: false }
});

const localPool = new Pool({
  connectionString: 'postgresql://postgres:fulebimojviT1985%@localhost:5432/ICE_ERP',
});

async function copyData() {
  const supabaseClient = await supabasePool.connect();
  const localClient = await localPool.connect();
  
  try {
    console.log('🗑️  Truncating local consolidated_bank_accounts table...');
    await localClient.query('TRUNCATE TABLE consolidated_bank_accounts CASCADE');
    console.log('✅ Local table truncated\n');
    
    console.log('📥 Fetching records from Supabase...');
    const result = await supabaseClient.query(`
      SELECT * FROM consolidated_bank_accounts ORDER BY id
    `);
    
    console.log(`Found ${result.rows.length} records`);
    
    if (result.rows.length === 0) {
      console.log('No data to copy');
      return;
    }
    
    console.log('Copying to local database...');
    let copied = 0;
    
    for (const row of result.rows) {
      await localClient.query(`
        INSERT INTO consolidated_bank_accounts (
          uuid, bank_account_uuid, raw_record_uuid, transaction_date,
          description, counteragent_uuid, project_uuid, financial_code_uuid,
          account_currency_uuid, account_currency_amount, nominal_currency_uuid,
          nominal_amount, created_at, updated_at
        ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14)
        ON CONFLICT (uuid) DO NOTHING
      `, [
        row.uuid, row.bank_account_uuid, row.raw_record_uuid, row.transaction_date,
        row.description, row.counteragent_uuid, row.project_uuid, row.financial_code_uuid,
        row.account_currency_uuid, row.account_currency_amount, row.nominal_currency_uuid,
        row.nominal_amount, row.created_at, row.updated_at
      ]);
      
      copied++;
      if (copied % 1000 === 0) {
        console.log(`  Copied ${copied}/${result.rows.length} records...`);
      }
    }
    
    console.log(`✅ Successfully copied ${copied} records to local database!`);
    
  } catch (error) {
    console.error('❌ Error:', error.message);
  } finally {
    supabaseClient.release();
    localClient.release();
    await supabasePool.end();
    await localPool.end();
  }
}

copyData();

File: copy_supabase_to_local.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Copy latest 1000 records from Supabase to LOCAL for testing
"""
import psycopg2
from dotenv import dotenv_values
from datetime import datetime

# Load environment
env = dotenv_values('.env.local')
local_url = env['DATABASE_URL'].split('?')[0]
# Remove query parameters from Supabase URL
remote_url = env['REMOTE_DATABASE_URL'].split('?')[0]

print("📥 Copying ALL records from Supabase to LOCAL...\n")

# Connect to both databases
remote_conn = psycopg2.connect(remote_url)
local_conn = psycopg2.connect(local_url)

remote_cur = remote_conn.cursor()
local_cur = local_conn.cursor()

# Find tables in Supabase
print("🔍 Finding BOG GEL raw tables in Supabase...")
remote_cur.execute("""
    SELECT tablename 
    FROM pg_tables 
    WHERE schemaname='public' AND tablename LIKE 'bog_gel_raw%'
    ORDER BY tablename DESC
    LIMIT 1
""")
remote_table = remote_cur.fetchone()

if not remote_table:
    print("❌ No BOG GEL raw tables found in Supabase")
    exit(1)

remote_table_name = remote_table[0]
print(f"   Found: {remote_table_name}")

# Check record count
remote_cur.execute(f"SELECT COUNT(*) FROM {remote_table_name}")
total_remote = remote_cur.fetchone()[0]
print(f"   Total records in Supabase: {total_remote}")

if total_remote == 0:
    print("❌ No records in Supabase table")
    exit(1)

# Find corresponding local table
local_cur.execute("""
    SELECT tablename 
    FROM pg_tables 
    WHERE schemaname='public' AND tablename LIKE 'bog_gel_raw%'
    ORDER BY tablename DESC
    LIMIT 1
""")
local_table = local_cur.fetchone()

if not local_table:
    print("❌ No BOG GEL raw tables found in LOCAL")
    print("💡 Create the table schema first")
    exit(1)

local_table_name = local_table[0]
print(f"   Local table: {local_table_name}")

# Get column names from BOTH tables and find common columns
remote_cur.execute(f"""
    SELECT column_name 
    FROM information_schema.columns 
    WHERE table_name = '{remote_table_name}'
    ORDER BY ordinal_position
""")
remote_columns = [col[0] for col in remote_cur.fetchall()]

local_cur.execute(f"""
    SELECT column_name 
    FROM information_schema.columns 
    WHERE table_name = '{local_table_name}'
    ORDER BY ordinal_position
""")
local_columns = [col[0] for col in local_cur.fetchall()]

# Find common columns
common_columns = [col for col in remote_columns if col in local_columns]

print(f"   Remote columns: {len(remote_columns)}")
print(f"   Local columns: {len(local_columns)}")
print(f"   Common columns: {len(common_columns)}")

# Fetch ALL records from Supabase (only common columns)
print(f"\n📤 Fetching ALL records from Supabase...")
columns_str = ', '.join(common_columns)
remote_cur.execute(f"""
    SELECT {columns_str}
    FROM {remote_table_name}
    ORDER BY created_at DESC
""")
records = remote_cur.fetchall()
print(f"   Fetched: {len(records)} records")

if not records:
    print("❌ No records fetched")
    exit(1)

# Clear local table
print(f"\n🗑️  Clearing local table {local_table_name}...")
local_cur.execute(f"TRUNCATE TABLE {local_table_name} CASCADE")
local_conn.commit()

# Insert into local table
print(f"📥 Inserting {len(records)} records into LOCAL...")
placeholders = ', '.join(['%s'] * len(common_columns))
insert_sql = f"INSERT INTO {local_table_name} ({columns_str}) VALUES ({placeholders})"

local_cur.executemany(insert_sql, records)
local_conn.commit()

# Verify
local_cur.execute(f"SELECT COUNT(*) FROM {local_table_name}")
local_count = local_cur.fetchone()[0]
print(f"✅ Successfully copied {local_count} records to LOCAL")

# Show batch info if column exists
if 'batch_id' in common_columns:
    local_cur.execute(f"""
        SELECT DISTINCT batch_id, COUNT(*) 
        FROM {local_table_name} 
        GROUP BY batch_id
        ORDER BY batch_id DESC
        LIMIT 5
    """)
    batches = local_cur.fetchall()
    if batches:
        print(f"\n📊 Records by batch (latest 5):")
        for batch, count in batches:
            print(f"   - Batch {batch}: {count} records")

# Close connections
remote_cur.close()
remote_conn.close()
local_cur.close()
local_conn.close()

print(f"\n✅ Ready for backparse testing!")
